_posts/2019-08-29-wfs3.md

```diff
@@ -0,0 +1,503 @@
+OpenAPI standard and JSON. This way, even if basic mechanisms are identical,
+concepts of `REQUEST`{:.inline} or `SERVICE`{:.inline} doesn't exist anymore.
+For example, instead of exploring a project with a `GetCapabilities`{:.inline}
+request, a JSON collection with layers' capabilities is retrieved with
+`http://qgisserver/wfs3/collections.json?MAP=/tmp/megaliths.qgz`{:.inline}.
+
+Such a collection is easily parsable in Python with the `json`{:.inline} module.
+
+{% capture summary %}`megaliths.py`{:.inline}{% endcapture %}
+{% capture details %}
+``` python
+import requests
+import json
+
+http = "http://qgisserver/wfs3/collections.json?MAP=/tmp/megaliths.qgz"
+request = requests.get(http)
+
+if request.status_code == 200:
+  doc = json.loads(request.text)
+
+  for layer in doc["collections"]:
+    if layer["title"] != "Megaliths":
+      continue
+
+    print("Title: ", layer["title"])
+    print("CRS: ", layer["extent"]["crs"])
+    print("Extent: ", layer["extent"]["spatial"])
+```
+{% endcapture %}
+<details>
+  <summary>{{ summary | markdownify | remove: '<p>' | remove: '</p>' }}</summary>
+  {{ details | markdownify }}
+</details>
+
+``` bash
+$ python megaliths.py
+Title: Megaliths
+CRS: http://www.opengis.net/def/crs/OGC/1.3/CRS84
+Extent: [[-5.150866731269053, -5.983856309208769, -1.048528989867108, 48.876193000478644]]
+```
+<br>
+
+Then, to discover `Megaliths`{:.inline} layer capabilities, nothing more easy
+than retrieving once again a JSON collection with
+`collections/Megaliths.json?`{:.inline}.
+
+{% capture summary %}`capabilities.py`{:.inline}{% endcapture %}
+{% capture details %}
+``` python
+import requests
+import json
+
+http = "http://qgisserver/wfs3/collections/Megaliths.json?MAP=/tmp/megaliths.qgz"
+request = requests.get(http)
+
+if request.status_code == 200:
+  doc = json.loads(request.text)
+
+  for link in doc["links"]:
+    if link["type"] != "application/json":
+      continue
+
+    print(link["href"])
+```
+{% endcapture %}
+<details>
+  <summary>{{ summary | markdownify | remove: '<p>' | remove: '</p>' }}</summary>
+  {{ details | markdownify }}
+</details>
+
+``` bash
+$ python capabilities.py
+http://qgisserver/wfs3/collections/Megaliths.json?MAP=/tmp/megaliths.qgz
+http://qgisserver/wfs3/collections/Megaliths/items.json?MAP=/tmp/megaliths.qgz
+````
+
+<br>
+
+As can be observed, there is a `collection/Megaliths/items.json`{:.inline}
+entry point. Actually, it allows to retrieve features for a dedicated layer,
+replacing the so called `GetFeature`{:.inline} request. In the same way,
+numerous parameters are available. However, there are some notable changes,
+especially about the CRS parameter. Indeed, the parameter is now named
+`bbox-crs`{:.inline} and is defined according to the content of the
+`crs`{:.inline} property specified within the JSON collection. Moreover, you can
+filter by adding `field_name=value`{:.inline} directly in the request instead
```

Comment:
maybe worth mentioning that there is support for partial matching with "*" (shell-like expansion)

Reply:
done